### PR TITLE
DOC: Clarify DataFrame column argument in API documentation

### LINKED ISSUE #39904: BUG: DataFrame construction with columns argument set to a MultiIndex creates an empty DataFrame.
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
>>> import pandas as pd
>>> from numpy import nan
>>>
>>>
>>> column_index = pd.MultiIndex.from_product([["sample1", "sample2"], [True], [True]],
...                                           names=["Name", "Sample_column", "Write"])
>>> table = pd.DataFrame(data={"sample1": [100.0, nan, 100.0, 10.0],
...                            "sample2": [80.0, 20.0, 100.0, 100.0]})
>>> table.columns = column_index
>>>
>>> table2 = pd.DataFrame(data={"sample1": [100.0, nan, 100.0, 10.0],
...                             "sample2": [80.0, 20.0, 100.0, 100.0]},
...                             columns=column_index)
>>> table
Name          sample1 sample2
Sample_column    True    True
Write            True    True
0               100.0    80.0
1                 NaN    20.0
2               100.0   100.0
3                10.0   100.0
>>> table2
Empty DataFrame
Columns: [(sample1, True, True), (sample2, True, True)]
Index:
```

#### Problem description

If a MultiIndex is used to set the columns using the columns argument from the constructor, the MultiIndex is set correctly but the whole dataframe becomes empty. If the columns index is applied manually after construction by using df.columns, the dataframe does not lose its data. I think this is either a bug, or a clear warning or error message should be given why the dataframe becomes empty.

#### Expected Output

I expect the constructor to return a non-empty dataframe with the MultiIndex applied to the columns. The API reference mentions that is can be an 'Index or array-like', but I don't know if a MultiIndex is allowed.

#### Output of ``pd.show_versions()``

<details>

>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit           : 7d32926db8f7541c356066dcadabf854487738de
python           : 3.8.5.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-58-generic
Version          : #64-Ubuntu SMP Wed Dec 9 08:16:25 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.2.2
numpy            : 1.20.1
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 44.0.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by DriesSchaumont**
- [x] closes #39904
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

### CODE REVIEW

**Comment by DriesSchaumont** [comment]
ping @rhshadrach

File: pandas/core/frame.py [hunk 1]
```diff
@@ -472,8 +472,12 @@ class DataFrame(NDFrame, OpsMixin):
         Index to use for resulting frame. Will default to RangeIndex if
         no indexing information part of input data and no index provided.
     columns : Index or array-like
-        Column labels to use for resulting frame. Will default to
-        RangeIndex (0, 1, 2, ..., n) if no column labels are provided.
+        Columns to select from data or column labels to use for resulting frame.
+        Will use the provided labels for the column index if data does not
+        include column labels, defaulting to RangeIndex(0, 1, 2, ..., n).
```
Comment (rhshadrach) [inline pandas/core/frame.py:477]: I think we can be a little more concise; what do you think of:

Column labels to use for resulting frame when data does not have them, defaulting to RangeIndex(0, 1, 2, ..., n). If data contains column labels, will perform column selection instead.

File: pandas/core/frame.py [hunk 2]
```diff
@@ -523,6 +527,16 @@ class DataFrame(NDFrame, OpsMixin):
     1  4  5  6
     2  7  8  9
 
+    >>> d = np.array([(1, 2, 3), (4, 5, 6), (7, 8, 9)],
```
Comment (rhshadrach) [inline pandas/core/frame.py:530]: Nice example - can you rename d to data

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR @DriesSchaumont

File: pandas/core/frame.py (hunk 1 above)
Comment (DriesSchaumont) [reply pandas/core/frame.py]: Yes, thats better! Adjusted

File: pandas/core/frame.py (hunk 2 above)
Comment (DriesSchaumont) [reply pandas/core/frame.py]: Done.

File: pandas/core/frame.py [hunk 3]
```diff
@@ -523,6 +524,16 @@ class DataFrame(NDFrame, OpsMixin):
     1  4  5  6
     2  7  8  9
 
+    >>> data = np.array([(1, 2, 3), (4, 5, 6), (7, 8, 9)],
```
Comment (jreback) [inline pandas/core/frame.py:527]: can you add a Description like above

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 3 above)
Comment (DriesSchaumont) [reply pandas/core/frame.py:529]: Done!

**Review by rhshadrach — APPROVED** [review]
lgtm

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@DriesSchaumont thanks a lot for your PR!

**Pull request merged**